gapis/service/types/types.go

```diff
@@ -0,0 +1,217 @@
+	}
+	type_map[i] = &typeData{t, rt}
+}
+
+func GetType(i uint64) (*Type, error) {
+	t, ok := type_map[i]
+	if !ok {
+		return nil, fmt.Errorf("Could not find type %v", i)
+	}
+	return t.tp, nil
+}
+
+func GetReflectedType(i uint64) (reflect.Type, error) {
+	t, ok := type_map[i]
+	if !ok {
+		return nil, fmt.Errorf("Could not find type %v", i)
+	}
+	return t.rt, nil
+}
+
+// GetTypeIndex returns the index of the type. This must be
+// kept in sync with api_types.go.tmpl
+func GetTypeIndex(i interface{}) (uint64, error) {
+	switch i := i.(type) {
+	case TypeProvider:
+		return i.GetTypeIndex(), nil
+	case bool:
+		return 1, nil
+	case memory.IntTy:
+		return 2, nil
+	case memory.UintTy:
+		return 3, nil
+	case memory.SizeTy:
+		return 4, nil
+	case memory.CharTy:
+		return 5, nil
+	case uint8:
+		return 6, nil
+	case int8:
+		return 7, nil
+	case uint16:
+		return 8, nil
+	case int16:
+		return 9, nil
+	case float32:
+		return 10, nil
+	case uint32:
+		return 11, nil
+	case int32:
+		return 12, nil
+	case float64:
+		return 13, nil
+	case uint64:
+		return 15, nil
+	case int64:
+		return 16, nil
+	case string:
+		return 17, nil
+	}
+	return 0, fmt.Errorf("Unknown type %T", i)
+}
+
+// TypeProvider is an interface for any type that
+// has a registered type.
+type TypeProvider interface {
+	GetTypeIndex() uint64
+}
+
+func init() {
+	AddType(1, &Type{
+		TypeId: 1,
+		Name:   "bool",
+		Ty: &Type_Pod{
+			pod.Type_bool,
+		},
+	}, reflect.TypeOf(bool(false)))
+	AddType(6, &Type{
+		TypeId: 6,
+		Name:   "uint8_t",
+		Ty: &Type_Pod{
+			pod.Type_uint8,
+		},
+	}, reflect.TypeOf(uint8(0)))
+	AddType(7, &Type{
+		TypeId: 7,
+		Name:   "int8_t",
+		Ty: &Type_Pod{
+			pod.Type_sint8,
+		},
+	}, reflect.TypeOf(int8(0)))
+	AddType(8, &Type{
+		TypeId: 8,
+		Name:   "uint16_t",
+		Ty: &Type_Pod{
+			pod.Type_uint16,
+		},
+	}, reflect.TypeOf(uint16(0)))
+	AddType(9, &Type{
+		TypeId: 9,
+		Name:   "int16_t",
+		Ty: &Type_Pod{
+			pod.Type_sint16,
+		},
+	}, reflect.TypeOf(int16(0)))
+	AddType(10, &Type{
+		TypeId: 10,
+		Name:   "float32",
+		Ty: &Type_Pod{
+			pod.Type_float32,
+		},
+	}, reflect.TypeOf(float32(0)))
+	AddType(11, &Type{
+		TypeId: 11,
+		Name:   "uint32_t",
+		Ty: &Type_Pod{
+			pod.Type_uint32,
+		},
+	}, reflect.TypeOf(uint32(0)))
+	AddType(12, &Type{
+		TypeId: 12,
+		Name:   "int32_t",
+		Ty: &Type_Pod{
+			pod.Type_sint32,
+		},
+	}, reflect.TypeOf(int32(0)))
+	AddType(13, &Type{
+		TypeId: 13,
+		Name:   "float64",
+		Ty: &Type_Pod{
+			pod.Type_float64,
+		},
+	}, reflect.TypeOf(float64(0)))
+	AddType(14, &Type{
+		TypeId: 14,
+		Name:   "uint64_t",
+		Ty: &Type_Pod{
+			pod.Type_uint64,
+		},
+	}, reflect.TypeOf(uint64(0)))
+	AddType(15, &Type{
+		TypeId: 15,
+		Name:   "int64_t",
+		Ty: &Type_Pod{
+			pod.Type_sint64,
+		},
+	}, reflect.TypeOf(int64(0)))
+	AddType(16, &Type{
+		TypeId: 16,
+		Name:   "string",
+		Ty: &Type_Pod{
+			pod.Type_string,
+		},
+	}, reflect.TypeOf(string("")))
+
+	AddType(2, &Type{
```

Comment:
Why aren't these types declaration between type index 1 and 6?

Reply:
I kept those types at the end because they are special. Their size changes based on ABI. Whereas the others have fixed sizes.